pandas/core/generic.py

```diff
@@ -3972,6 +3972,16 @@ def asof(self, where, subset=None):
             where = Index(where) if is_list else Index([where])
 
         nulls = self.isnull() if is_series else self[subset].isnull().any(1)
+        if nulls.all():
+            if is_series:
+                return pd.Series(np.nan, index=where)
+            else:
+                if is_list:
+                    return pd.DataFrame(np.nan, index=where,
+                                        columns=self.columns)
+                else:
+                    return pd.Series(np.nan, index=self.columns, name=where[0])
+
```

Comment:
hey @jreback , I made a small simplification, pls check if that's ok... if it's ok, now I think everything is good to go